File: test/echo-test/go-client/echo_client.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"net"
	"os"
	"time"
)

func main() {
	ip := "127.0.0.1"
	port := 65432

	// Connect to the server
	conn, err := net.Dial("tcp", fmt.Sprintf("%s: %d", ip, port))
	fmt.Printf("Connecting to %s:%d ... \n", ip, port)
	if err != nil {
		log.Println(err)
		return
	}
	defer conn.Close()

	fmt.Printf("reading... \n")

	for {
		//send keyboard input from a user to the server
		// read in input from stdin
		reader := bufio.NewReader(os.Stdin)
		text, _ := reader.ReadString('\n')

		// Send a message
		fmt.Fprintf(conn, text)

		// Receive and print the response
		response, err := bufio.NewReader(conn).ReadString('\n')
		if err != nil {
			fmt.Println(err)
			return
		}
		fmt.Print("response: ", response)

		time.Sleep(2 * time.Second)

		defer func() {
			if r := recover(); r != nil {
				log.Println("Error:", r)
			}
		}()
	}
}

File: test/echo-test/rust-client/echo-client.rs
use std::io::{self, Read, Write};
use std::{thread, time::Duration};
use std::net::TcpStream;

fn main() -> io::Result<()> {
    let ip = "127.0.0.1";
    let port = 65432;

    // Connect to the server
    let mut stream = TcpStream::connect(format!("{}:{}", ip, port))?;
    //if stream is ok
    println!("Connected to {}:{}", ip, port);

    println!("reading...");
    // write a while true loop
    loop {
      //send keyboard input from a user to the server
      let mut input = String::new();
      io::stdin().read_line(&mut input)?;
      
      // Send the input to the server
      stream.write_all(input.as_bytes())?;

      //read the response from the server
      let mut buffer = [0; 1024];
      stream.read(&mut buffer)?;
      let response = String::from_utf8_lossy(&buffer[..]);
      println!("response: {}", response);

      thread::sleep(Duration::from_millis(500));
    }
}